### DISCUSSION TITLE
Sharper text rendering

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre-style-spec#174] [Ideas]
Our current SDF-based text rendering stack leads inevitably to rounded edges on text which are an artifact from the bilinear interpolation which is used to turn a raster SDF to pixel colors.

To get sharper text rendering, one approach could be to double the resolution of all the SDF raster data as @bdon has proposed. This might be a very direct way to get sharper text.

An alternative approach might be to upload spline segments to the GPU for every glyph and compute the distance to the glyph on the GPU. It is more computationally intensive than our current raster SDF texture approach, but the advantage would be that it creates much sharper text.

There is an OpenGL ES2 implementation of this, it is called https://github.com/behdad/glyphy and is written by Behdad, the author of HarfBuzz, who is a text shaping and rendering expert. Anyway, might be worth exploring, our text looks bad I find when comparing it to Google Maps...

**Comment by HarelM** [comment]
While I can't argue that the text may need improvements, I'm not sure I understand what will be the needed change in the spec...?

**Reply by wipfli** [reply]
The font API would change. But do we actually document the font API in the style spec?

**Reply by 1ec5** [reply]
Yes, in passing, and there are deployments of MapLibre SDKs out there that depend on this behavior:

https://maplibre.org/maplibre-style-spec/root/#glyphs

**Comment by bdon** [comment]
I opened a GL JS issue here https://github.com/maplibre/maplibre-gl-js/issues/2990 which proposes improvements that do not affect the style spec
